core/src/test/java/org/apache/calcite/rex/RexProgramTest.java

```diff
@@ -801,6 +804,17 @@ private void checkExponentialCnf(int n) {
                         and(eRef,
                             or(fRef,
                                and(gRef, or(trueLiteral, falseLiteral)))))))));
+
```

Comment:
Did you make sure the test fails with the original `HashMap` implementation?

Reply:
> Did you make sure the test fails with the original `HashMap` implementation?

Hi @asolimando ,I tested with tpc-h q19.sql,and I printed some information before and after calling the method RexUtil.pullFactors,at the same time, in my program, I monitor all tpch sql plans, and if the plan changes, my program will sense.

![image](https://github.com/apache/calcite/assets/8057451/262e54c4-129e-450f-a3ad-ca8bb7c71e03)

In the picture above,simplifiedCondition is the RexNode before entering the RexUtil.pullFactors method,factors.values() represents the result of calling RexUtil.CnfHelper#commonFactors , after pullFactors is the RexNode after entering the RexUtil.pullFactors method.As you can see,RexNode might be different after entering the method RexUtil.pullFactors.

After I changed the HashMap of RexUtil.CnfHelper#commonFactors to LinkedHashMap, the above problem never appeared again.

Reply:
> We should find a way to prevent regressions on this, and bake them into a test that can run within Calcite, we cannot rely too much on such manual checks as people come and go, while Calcite tests stay.
> 
> I wonder then what's the extra part you added to the test is actually checking if not predicate order.

Hi @asolimando ,the single test I added is the case shown in my picture above,I improved the content of the single test I added.

Reply:
When I try the `testCnfExponential` test with `HashMap` and `LinkedHashMap`, it passes in both cases, am I missing something?

Reply:
Mmm, I think I have to execute the test until it fails (the order is non-deterministic without your fix), sometimes it can be OK but it's not guaranteed.

I don't see a way we can improve this test further (e.g., making it fail with `HashMap` consistently), because of the nature of the change, but I think it's already something that the test will become flaky if in the future we break the determinism you just introduced thanks to the `LinkedHashMap`.

In summary, I am OK with the test, I don't see how it can be (easily) improved further.

Reply:
> When I try the `testCnfExponential` test with `HashMap` and `LinkedHashMap`, it passes in both cases, am I missing something?

Hi @asolimando , testCnfExponential method  doesn't call RexUtil.CnfHelper#commonFactors, and the out-of-order situation only occurs in some cases. In my PR test case, the following operation reproduces 100% of the out-of-order situation:
1. use HashMap
2. change the exprected result of my test case to :
```sql
"AND(?0.c, SEARCH(?0.i, Sarg['AIR':CHAR(7), 'AIR REG']:CHAR(7)), =(?0.a, ?0.b), OR(AND(=(?0.j, 'Brand#12'), >=(?0.h, 1), <=(?0.h, 11), SEARCH(?0.k, Sarg['SM BOX':CHAR(7), 'SM CASE', 'SM PACK', 'SM PKG':CHAR(7)]:CHAR(7))), AND(=(?0.j, 'Brand#13'), >=(?0.h, 10), <=(?0.h, 20), SEARCH(?0.k, Sarg['MED BOX':CHAR(8), 'MED CASE', 'MED PACK', 'MED PKG':CHAR(8)]:CHAR(8))), AND(=(?0.j, 'Brand#14'), >=(?0.h, 20), <=(?0.h, 30), SEARCH(?0.k, Sarg['LG BOX':CHAR(7), 'LG CASE', 'LG PACK', 'LG PKG':CHAR(7)]:CHAR(7)))))"
```
3.  Separate Run and Debug RexProgramTest#testPullFactors, directly run single test can pass, Debug single test will give an error.

I see the underlying RexNode HashCode is different, currently I am investigating why run and Debug, For the same RexNode, the hashCode is different. In HashMap,the hashcode affects the order in which the values method returns results.

Change the HashMap to LinkedHashMap,the values method returns results in a deterministic order,it no longer has the possibility of being out of order.What do you think

Reply:
Sorry I picked the wrong test name, the correct one is `RexProgramTest#testPullFactors` as you mention, running it twice (after replacing `LinkedHashMap` to `HashMap`) gave a difference, so the test is effective:

```
RexUtil.pullFactors(rexBuilder, OR(AND(=(?0.a, ?0.b), =(?0.j, 'Brand#12'), >=(?0.h, 1), <=(?0.h, 11), ?0.c, SEARCH(?0.i, Sarg['AIR':CHAR(7), 'AIR REG']:CHAR(7)), SEARCH(?0.k, Sarg['SM BOX':CHAR(7), 'SM CASE', 'SM PACK', 'SM PKG':CHAR(7)]:CHAR(7))), AND(=(?0.a, ?0.b), =(?0.j, 'Brand#13'), >=(?0.h, 10), <=(?0.h, 20), ?0.c, SEARCH(?0.i, Sarg['AIR':CHAR(7), 'AIR REG']:CHAR(7)), SEARCH(?0.k, Sarg['MED BOX':CHAR(8), 'MED CASE', 'MED PACK', 'MED PKG':CHAR(8)]:CHAR(8))), AND(=(?0.a, ?0.b), =(?0.j, 'Brand#14'), >=(?0.h, 20), <=(?0.h, 30), ?0.c, SEARCH(?0.i, Sarg['AIR':CHAR(7), 'AIR REG']:CHAR(7)), SEARCH(?0.k, Sarg['LG BOX':CHAR(7), 'LG CASE', 'LG PACK', 'LG PKG':CHAR(7)]:CHAR(7)))))
Expected: with toString() "AND(=(?0.a, ?0.b), ?0.c, SEARCH(?0.i, Sarg['AIR':CHAR(7), 'AIR REG']:CHAR(7)), OR(AND(=(?0.j, 'Brand#12'), >=(?0.h, 1), <=(?0.h, 11), SEARCH(?0.k, Sarg['SM BOX':CHAR(7), 'SM CASE', 'SM PACK', 'SM PKG':CHAR(7)]:CHAR(7))), AND(=(?0.j, 'Brand#13'), >=(?0.h, 10), <=(?0.h, 20), SEARCH(?0.k, Sarg['MED BOX':CHAR(8), 'MED CASE', 'MED PACK', 'MED PKG':CHAR(8)]:CHAR(8))), AND(=(?0.j, 'Brand#14'), >=(?0.h, 20), <=(?0.h, 30), SEARCH(?0.k, Sarg['LG BOX':CHAR(7), 'LG CASE', 'LG PACK', 'LG PKG':CHAR(7)]:CHAR(7)))))"
     but: toString() was "AND(SEARCH(?0.i, Sarg['AIR':CHAR(7), 'AIR REG']:CHAR(7)), =(?0.a, ?0.b), ?0.c, OR(AND(=(?0.j, 'Brand#12'), >=(?0.h, 1), <=(?0.h, 11), SEARCH(?0.k, Sarg['SM BOX':CHAR(7), 'SM CASE', 'SM PACK', 'SM PKG':CHAR(7)]:CHAR(7))), AND(=(?0.j, 'Brand#13'), >=(?0.h, 10), <=(?0.h, 20), SEARCH(?0.k, Sarg['MED BOX':CHAR(8), 'MED CASE', 'MED PACK', 'MED PKG':CHAR(8)]:CHAR(8))), AND(=(?0.j, 'Brand#14'), >=(?0.h, 20), <=(?0.h, 30), SEARCH(?0.k, Sarg['LG BOX':CHAR(7), 'LG CASE', 'LG PACK', 'LG PKG':CHAR(7)]:CHAR(7)))))"
java.lang.AssertionError: RexUtil.pullFactors(rexBuilder, OR(AND(=(?0.a, ?0.b), =(?0.j, 'Brand#12'), >=(?0.h, 1), <=(?0.h, 11), ?0.c, SEARCH(?0.i, Sarg['AIR':CHAR(7), 'AIR REG']:CHAR(7)), SEARCH(?0.k, Sarg['SM BOX':CHAR(7), 'SM CASE', 'SM PACK', 'SM PKG':CHAR(7)]:CHAR(7))), AND(=(?0.a, ?0.b), =(?0.j, 'Brand#13'), >=(?0.h, 10), <=(?0.h, 20), ?0.c, SEARCH(?0.i, Sarg['AIR':CHAR(7), 'AIR REG']:CHAR(7)), SEARCH(?0.k, Sarg['MED BOX':CHAR(8), 'MED CASE', 'MED PACK', 'MED PKG':CHAR(8)]:CHAR(8))), AND(=(?0.a, ?0.b), =(?0.j, 'Brand#14'), >=(?0.h, 20), <=(?0.h, 30), ?0.c, SEARCH(?0.i, Sarg['AIR':CHAR(7), 'AIR REG']:CHAR(7)), SEARCH(?0.k, Sarg['LG BOX':CHAR(7), 'LG CASE', 'LG PACK', 'LG PKG':CHAR(7)]:CHAR(7)))))
Expected: with toString() "AND(=(?0.a, ?0.b), ?0.c, SEARCH(?0.i, Sarg['AIR':CHAR(7), 'AIR REG']:CHAR(7)), OR(AND(=(?0.j, 'Brand#12'), >=(?0.h, 1), <=(?0.h, 11), SEARCH(?0.k, Sarg['SM BOX':CHAR(7), 'SM CASE', 'SM PACK', 'SM PKG':CHAR(7)]:CHAR(7))), AND(=(?0.j, 'Brand#13'), >=(?0.h, 10), <=(?0.h, 20), SEARCH(?0.k, Sarg['MED BOX':CHAR(8), 'MED CASE', 'MED PACK', 'MED PKG':CHAR(8)]:CHAR(8))), AND(=(?0.j, 'Brand#14'), >=(?0.h, 20), <=(?0.h, 30), SEARCH(?0.k, Sarg['LG BOX':CHAR(7), 'LG CASE', 'LG PACK', 'LG PKG':CHAR(7)]:CHAR(7)))))"
     but: toString() was "AND(SEARCH(?0.i, Sarg['AIR':CHAR(7), 'AIR REG']:CHAR(7)), =(?0.a, ?0.b), ?0.c, OR(AND(=(?0.j, 'Brand#12'), >=(?0.h, 1), <=(?0.h, 11), SEARCH(?0.k, Sarg['SM BOX':CHAR(7), 'SM CASE', 'SM PACK', 'SM PKG':CHAR(7)]:CHAR(7))), AND(=(?0.j, 'Brand#13'), >=(?0.h, 10), <=(?0.h, 20), SEARCH(?0.k, Sarg['MED BOX':CHAR(8), 'MED CASE', 'MED PACK', 'MED PKG':CHAR(8)]:CHAR(8))), AND(=(?0.j, 'Brand#14'), >=(?0.h, 20), <=(?0.h, 30), SEARCH(?0.k, Sarg['LG BOX':CHAR(7), 'LG CASE', 'LG PACK', 'LG PKG':CHAR(7)]:CHAR(7)))))"
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
	at org.apache.calcite.rex.RexProgramTestBase.checkPullFactors(RexProgramTestBase.java:63)
	at org.apache.calcite.rex.RexProgramTest.testPullFactors(RexProgramTest.java:813)
[...]
```

As I was saying, it seems difficult to make it fail more consistently, because that's exactly the point of your PR, so I think this test, as-is, is good enough, because:

- it fails at times, so if we break the determinism at least we will have a flaky test, which will force people to look into what happened
- it documents explicitly that we want a deterministic output for that method, if the test becomes flaky people will easily figure out why from your comments (both in the unit-test and in the `commonFactors` method)

Another possible addition is to make `commonFactors` return `LinkedHashMap<RexNode, RexNode>` instead of `Map<RexNode, RexNode>`, to stress that we really want a deterministic order (and therefore, `Map<>` would not be good). This is one of the few cases where it makes sense to use a concrete implementation in place of an interface. 

The `commonFactors` method is `private static` and is called in a single place, it wouldn't cause much troubles changing its return type, we can consider this but I don't feel strongly about it, it's just a stronger way to document our expectations on the method other than a comment.

WDYT?